### PR TITLE
cmake: kconfig: Create a single source of truth

### PR DESCRIPTION
**Pull request by sigvartmh**
Add a text file which is read by CMake as a list for setting
environmental variables required by Kconfig. This is to be consumed by
third-party applications/scripts which wants to use Kconfig and need to
know which environmental variables we require to be set.

This is coupled with https://github.com/zephyrproject-rtos/ci-tools/pull/27

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

File: cmake/kconfig.cmake [hunk 1]
```diff
@@ -23,19 +23,17 @@ if(OVERLAY_CONFIG)
...
-set(ENV{ARCH_DIR}   ${ARCH_DIR})
-set(ENV{GENERATED_DTS_BOARD_CONF} ${GENERATED_DTS_BOARD_CONF})
+set(srctree ${ZEPHYR_BASE})
+set(KCONFIG_CONFIG ${DOTCONFIG})
+foreach(env_var ${KCONFIG_ENV_VARS_AS_LIST})
+  set(ENV{${env_var}} ${${env_var}})
+endforeach(env_var)
 
 add_custom_target(
   menuconfig
```
Comment (SebastianBoe) [inline cmake/kconfig.cmake]: Could we use this to get rid of the duplication below?

File: cmake/kconfig.cmake [hunk 2]
```diff
@@ -23,19 +23,17 @@ if(OVERLAY_CONFIG)
...
 endif()
 
-set(ENV{srctree}            ${ZEPHYR_BASE})
-set(ENV{KERNELVERSION}      ${KERNELVERSION})
-set(ENV{KCONFIG_CONFIG}     ${DOTCONFIG})
-set(ENV{PYTHON_EXECUTABLE} ${PYTHON_EXECUTABLE})
+# Read environmental variables which needs to be set for Kconfig
+# Idealy the variable name and environmental variable name should be the same
+file(STRINGS ${CMAKE_CURRENT_LIST_DIR}/kconfig_env_vars.txt KCONFIG_ENV_VARS_AS_LIST)
+print(KCONFIG_ENV_VARS_AS_LIST)
```
Comment (SebastianBoe) [inline cmake/kconfig.cmake]: The print should be omitted

File: cmake/kconfig.cmake [hunk 3]
```diff
@@ -23,19 +23,17 @@ if(OVERLAY_CONFIG)
   string(REPLACE " " ";" OVERLAY_CONFIG_AS_LIST "${OVERLAY_CONFIG}")
 endif()
 
-set(ENV{srctree}            ${ZEPHYR_BASE})
-set(ENV{KERNELVERSION}      ${KERNELVERSION})
-set(ENV{KCONFIG_CONFIG}     ${DOTCONFIG})
-set(ENV{PYTHON_EXECUTABLE} ${PYTHON_EXECUTABLE})
+# Read environmental variables which needs to be set for Kconfig
+# Idealy the variable name and environmental variable name should be the same
```
Comment (SebastianBoe) [inline cmake/kconfig.cmake]: s/Idealy/ideally/

**Review by SebastianBoe — CHANGES_REQUESTED** [review]
This looks good, it will allow third-parties to know which env vars they need to set to process Kconfig in a valid way.

A few minor nitpicks.

File: cmake/kconfig.cmake (hunk 2 above)
Comment (sigvartmh) [inline cmake/kconfig.cmake]:
```suggestion
```

File: cmake/kconfig.cmake (hunk 3 above)
Comment (sigvartmh) [inline cmake/kconfig.cmake]:
```suggestion
# Ideally the variable name and environmental variable name should be the same
```

File: cmake/kconfig.cmake (hunk 1 above)
Comment (sigvartmh) [reply cmake/kconfig.cmake]: I think so I'll try and add it if it works

**Review by SebastianBoe — APPROVED** [review]

**Comment by mbolivar** [comment]
Approved for quite a while - @sigvartmh, @SebastianBoe, still relevant?

**Comment by sigvartmh** [comment]
as https://github.com/zephyrproject-rtos/ci-tools/pull/27 is probably not needed I'm not sure how welcome this change is. If you think it is an improvement I can rebase it but if it adds unnecessary complexity I guess I'll just close it.

**Comment by SebastianBoe** [comment]
We can close and then re-open when needed.

**Pull request closed without merging**